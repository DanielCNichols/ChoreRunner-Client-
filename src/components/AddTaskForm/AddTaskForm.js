import React, { useState } from 'react';
import {
  Input,
  FormElement,
  Label,
  Fieldset,
  Legend,
  Select,
} from '../Form/Form';
import ApiService from '../../services/api-service';
import s from './AddTaskForm.module.css';

export default function AddTask({ members, addTask, toggleAdd, householdId }) {
  const [inputs, setInputs] = useState({
    title: '',
    points: '',
    member_id: '',
    household_id: householdId,
  });

  const [error, setError] = useState({
    title: null,
    points: null,
    member: null,
    server: null,
  });

  function handleInputChange(ev) {
    ev.persist();
    setInputs(inputs => ({
      ...inputs,
      [ev.target.name]: ev.target.value,
    }));
  }

  function resetErrors() {
    setError({
      title: null,
      points: null,
      member: null,
      server: null,
    });
  }

  function validateInput() {
    let errors = {};
    let { title, points, member_id } = inputs;

    if (!title) {
      errors.title = 'Title is required';
    }

    if (!points) {
      errors.points = 'You must assign a point value';
    }

    if (!member_id) {
      errors.member = 'You must assign the task to a member';
    }

    return errors;
  }

  async function handleSubmit(ev) {
    ev.preventDefault();
    resetErrors();
    try {
      let error = validateInput();

      if (error.title || error.member || error.points) {
        return setError(error);
      } else {
        let newTask = await ApiService.addTask(inputs);

        addTask(newTask);
        toggleAdd();
      }
    } catch (serverError) {
      setError({ ...error, server: serverError });
    }
  }

  return (
    <form className={s.addTaskForm} onSubmit={handleSubmit}>
      <Fieldset>
        <Legend>Assign Tasks</Legend>
        <FormElement className={s.formElement}>
          <Label htmlFor="taskName">Task Name</Label>
          {error.title && (
            <div role="alert" className={s.error}>
              <span>{error.title}</span>
            </div>
          )}
          <Input
            value={inputs.title}
            onChange={handleInputChange}
            placeholder="Clean Room"
            id="taskName"
            type="text"
            name="title"
          />
        </FormElement>

        <FormElement className={s.formElement}>
          <Label htmlFor="assignee">Assign to</Label>
          {error.member && (
            <div role="alert" className={s.error}>
              <span>{error.member}</span>
            </div>
          )}
          <Select
            value={inputs.member_id}
            onChange={handleInputChange}
            name="member_id"
            id="assignee"
            className="select-css"
          >
            <option value="">Please select a member</option>
            {members.map(member => {
              return (
                <option key={member.id} value={member.id}>
                  {member.name}
                </option>
              );
            })}
          </Select>
        </FormElement>

        <FormElement className={s.formElement}>
          <Label htmlFor="points">Points</Label>
          {error.points && (
            <div role="alert" className={s.error}>
              <span>{error.points}</span>
            </div>
          )}
          <Input
            value={inputs.points}
            onChange={handleInputChange}
            placeholder="2"
            type="number"
            min="1"
            max="100"
            id="points"
            name="points"
          />
        </FormElement>

        {error.server && (
          <div role="alert" className={s.error}>
            <p>{error.server.message}</p>
          </div>
        )}

        <div className={s.formButtons}>
          <button type="button" onClick={toggleAdd}>
            Cancel
          </button>
          <button type="submit">Assign Task</button>
        </div>
      </Fieldset>
    </form>
  );
}
