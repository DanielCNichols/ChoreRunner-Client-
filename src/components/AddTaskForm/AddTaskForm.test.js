import React from 'react';
import ReactDOM from 'react-dom';
import AddTaskForm from './AddTaskForm';
import renderer from 'react-test-renderer';

describe('AddTaskForm tests', () => {
  const members = [
    {
      id: 1,
      name: 'daniel',
    },
  ];

  const addTask = () => {};
  const toggleAdd = () => {};

  it('renders AddTask without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(
      <AddTaskForm members={members} toggleAdd={toggleAdd} addTask={addTask} />,
      div
    );
    ReactDOM.unmountComponentAtNode(div);
  });

  it('renders AddTask as expected', () => {
    const tree = renderer
      .create(
        <AddTaskForm
          members={members}
          toggleAdd={toggleAdd}
          addTask={addTask}
        />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
