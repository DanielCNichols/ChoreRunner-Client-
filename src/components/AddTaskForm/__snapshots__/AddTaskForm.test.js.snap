// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddTaskForm tests renders AddTask as expected 1`] = `
<form
  className="addTaskForm"
  onSubmit={[Function]}
>
  <fieldset
    className="Fieldset"
  >
    <legend
      className="Legend"
    >
      Assign Tasks
    </legend>
    <div
      className="FormElement formElement"
    >
      <label
        className="Label"
        htmlFor="taskName"
      >
        Task Name
      </label>
      <input
        className="Input"
        id="taskName"
        name="title"
        onChange={[Function]}
        placeholder="Clean Room"
        type="text"
        value=""
      />
    </div>
    <div
      className="FormElement formElement"
    >
      <label
        className="Label"
        htmlFor="assignee"
      >
        Assign to
      </label>
      <select
        className="Select select-css"
        id="assignee"
        name="member_id"
        onChange={[Function]}
        value=""
      >
        <option
          value=""
        >
          Please select a member
        </option>
        <option
          value={1}
        >
          daniel
        </option>
      </select>
    </div>
    <div
      className="FormElement formElement"
    >
      <label
        className="Label"
        htmlFor="points"
      >
        Points
      </label>
      <input
        className="Input"
        id="points"
        max="100"
        min="1"
        name="points"
        onChange={[Function]}
        placeholder="2"
        type="number"
        value=""
      />
    </div>
    <div
      className="formButtons"
    >
      <button
        onClick={[Function]}
        type="button"
      >
        Cancel
      </button>
      <button
        type="submit"
      >
        Assign Task
      </button>
    </div>
  </fieldset>
</form>
`;
